Switch indexeddb backend to sqlite and improve IPC messaging

arihant2math opened the pull request:
- Use sqlite instead of heed. (one indexed database = one sqlite database)
- Implement the backend for indexes
- Use keyranges where needed (as specified by the spec)
- Implement `getKey`
- Fix channel error messaging (led to a bunch of changes to how async requests are handled)

Note: `components/net/indexeddb/engines/sqlite/serialize.rs` is unused; I can delete it if needed.

Testing: Switching to sqlite eliminated many panics (exposing some new failures).
Fixes: #38040

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

arihant2math commented:
Note to self: use git-cherrypick next time.

janvarga commented on components/net/indexeddb/engines/sqlite/database_model.rs:
```diff
@@ -0,0 +1,22 @@
+#[sea_orm(table_name = "database")]
+pub struct Model {
+    #[sea_orm(primary_key, auto_increment = false)]
+    pub name: String,
```
I'm not sure how this maps to the SQLite column type, but make sure you're not using TEXT. SQLite defaults to UTF-8 encoding, so if you store a string that came from JavaScript and then read it back, you can lose information, especially with things like lone surrogates.

Even though it may seem unusual, database names, object store names, index names, etc., should be stored as BLOBs in the table to preserve exact byte representation. For reference, see [bug 1300205](https://bugzilla.mozilla.org/show_bug.cgi?id=1300205).

arihant2math replied:
Good point.

According to https://www.sea-ql.org/SeaORM/docs/next/generate-entity/entity-structure/#column-type, String maps to varchar, I doubt that falls into a different category.

Either
```
PRAGMA encoding = 'UTF-16';
``` 
or opening the database in utf-16 format should solve this. At some point ruqlite will be added as a dependency because I need to be able to run vacuuming. I suppose at that point we could modify the database creation code to create with UTF-16.

There should probably be a wpt test for this, I'll add one if there isn't.

arihant2math replied:
Fixed by 6157f8cde2870af129150301a4615659308e165e

jdm commented:
> @jdm I think the pr is ready for review when you get the time.

I really appreciate the work you're doing here, but batching multiple PRs together into one PR with 26 commits is not going to work for me. Incremental work that can exist in its own PR is easier to 1) review, 2) understand, and 3) identify regressions. I would really appreciate if you could extract this work into separate PRs.

arihant2math commented:
I can probably branch off part of key ranges and the cmp method, but the bulk of the work is tightly coupled (I rewrote the messaging system when I rewrote the backend system to reduce the amount of work needed).

arihant2math committed:
Port to sqlite

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

arihant2math commented:
@jdm after #38278, do you think you could take another look at this?

jdm commented on components/script/dom/idbrequest.rs:
```diff
@@ -68,6 +122,40 @@ impl RequestListener {
+                    // no-op
+                },
+                IdbResult::Error(_err) => {
+                    // FIXME:(arihant2math) duplicated
```
Can we do this error work in a closure here to remove the duplication?

jdm commented on components/script/dom/idbtransaction.rs:
```diff
@@ -229,6 +229,7 @@ impl IDBTransactionMethods<crate::DomTypeHolder> for IDBTransaction {
         // returns the same IDBObjectStore instance.
         let mut store_handles = self.store_handles.borrow_mut();
         let store = store_handles.entry(name.to_string()).or_insert_with(|| {
+            // get key path
```
I don't know what this comment means.

arihant2math replied:
Fixed the comment, it was meant to be a todo.

jdm commented on components/shared/net/indexeddb_thread.rs:
```diff
@@ -200,33 +205,47 @@ impl IndexedDBKeyRange {
-    ),
+    PutItem {
+        sender: IpcSender<DbResult<PutItemResult>>,
+        key: IndexedDBKeyType, // Key
```
I don't think this comment adds any value.

jdm commented on components/shared/net/indexeddb_thread.rs:
```diff
@@ -237,50 +256,84 @@ pub enum AsyncOperation {
     /// Upgrades the version of the database
     UpgradeVersion(
-        IpcSender<Result<u64, ()>>,
+        IpcSender<Option<u64>>,
```
What does it mean if a None value is sent?

jdm commented on components/shared/net/indexeddb_thread.rs:
```diff
@@ -237,50 +256,84 @@ pub enum AsyncOperation {
     /// Checks if an object store has a key generator, used in e.g. Put
     HasKeyGenerator(
-        IpcSender<bool>,
+        IpcSender<Option<bool>>,
```
What does it mean if a None value is sent?

jdm commented on components/shared/net/indexeddb_thread.rs:
```diff
@@ -237,50 +256,84 @@ pub enum AsyncOperation {
+        String, // Database
+        String, // Store
+    ),
+    /// Gets an object stores key path
```
```suggestion
    /// Gets an object store's key path
```

jdm commented on components/shared/net/indexeddb_thread.rs:
```diff
@@ -237,50 +256,84 @@ pub enum AsyncOperation {
+    ),
+    /// Gets an object stores key path
+    KeyPath(
+        IpcSender<Option<Option<KeyPath>>>,
```
What does a None value mean? What does a Some(None) value mean?

jdm commented on components/net/indexeddb/engines/sqlite/serialize.rs:
```diff
@@ -0,0 +1,360 @@
+
+/// A module for encoding and decoding various IndexedDBKeyType types according to the basic strategy:
+///
+/// - Numbers:  0x10 + 8-byte sortable float
+/// - Dates:    0x20 + 8-byte sortable float
+/// - Strings:  0x30 + encoded chars + 0 terminator
+/// - Binaries: 0x40 + encoded octets + 0 terminator
+/// - Arrays:   0x50 + encoded items + 0 terminator
```
Can we do:
```rust
enum SerializedValuePrefix {
    Number = 0x10,
    Date = 0x20,
    String = 0x30,
    Binary = 0x40,
    Array = 0x50,
}
```
and then refer to those values instead of using magic constants in this file?

jdm commented on components/net/indexeddb/engines/sqlite/serialize.rs:
```diff
@@ -0,0 +1,360 @@
+}
+
+/// Encode a date with prefix 0x20 and terminator 0x00.
+pub fn encode_date(data: &[u8]) -> Vec<u8> {
```
I would find it helpful if we could document the intention of these routines. Are they following some kind of specification, or using another implementation as a model?

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+mod object_data_model;
+mod object_store_index_model;
+mod object_store_model;
+// mod serialize;
+mod serialize {
```
What's the difference between the serialization in this file vs. the one in the script crate?

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+    condition
+}
+
+macro_rules! err {
+    ($e:expr) => {
+        Err(format!("{:?}", $e))
+    };
+}
```
Same comments as the other err! macro in this PR. I don't think it provides enough value to keep it.

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+    }
+
+    async fn get_connection(path: &Path) -> Result<DatabaseConnection, DbErr> {
+        Database::connect(format!("sqlite://{}", path.display())).await
```
It looks odd to await here, since we immediately return and the caller calls await. Can we remove this await?

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+        db_info: &IndexedDBDescription,
+        version: u64,
+    ) -> Result<DatabaseConnection, DbErr> {
+        macro_rules! create_table {
```
Why is this a macro instead of a function or a closure?

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+            let model = object_store_model::ActiveModel {
+                id: NotSet,
+                name: Set(store_name.to_string()),
+                key_path: Set(key_path.map(|v| bincode::serialize(&v).unwrap())),
```
Should we use this crate's serialize module for this instead?

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+        HANDLE.block_on(async {
+            self.connection.close().await?;
+            if self.db_path.exists() {
+                std::fs::remove_dir_all(self.db_path.parent().unwrap()).unwrap();
```
Can we return errors instead of unwrapping here?

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+                    Some(object_store) => object_store,
+                    None => {
+                        // TODO: This is also kinda wrong, but atleast we don't panic.
+                        tx.send(None).unwrap_or(());
```
```suggestion
                        let _ = tx.send(None);
```

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+                    .filter(object_store_model::Column::Name.eq(request.store_name.to_string()))
+                    .one(&conn)
+                    .await
+                    .unwrap()
```
What are we unwrapping here? Can we return an error instead?

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+                                )
+                                .one(&conn)
+                                .await
+                                .unwrap() // TODO: handle
```
What does it take to handle this now instead of leaving a TODO?

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+            let db_info = database_model::Entity::find()
+                .one(&self.connection)
+                .await
+                .unwrap()
+                .unwrap();
```
Can we avoid this?

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+                .unwrap()
+                .unwrap();
+            let mut db_info = db_info.into_active_model();
+            db_info.version = Set(version as i32);
```
I'm nervous that our public API takes values that are unrepresentable in the actual implementation. What can we do about that?

arihant2math replied:
Good point, maybe just using an i64 should solve the issue.

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+            let db_info = database_model::Entity::find()
+                .one(&self.connection)
+                .await
+                .unwrap()
+                .unwrap();
```
Can we avoid this?

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+                id: Default::default(),
+                object_store_id: Set(object_store.id),
+                name: Set(index_name),
+                key_path: Set(bincode::serialize(&key_path).unwrap()),
```
Should we use the serialize module instead?

arihant2math replied:
I don't think so, the serialize module is just to ensure that keys get properly sorted and so that key ranges work.

jdm commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,513 @@
+            {
+                model
+                    .key_path
+                    .map(|key_path| bincode::deserialize(&key_path).unwrap())
```
Should we use the serialize module instead?

arihant2math replied:
See above

jdm reviewed (changes requested):
I've read through the PR! I'm still unhappy that it includes new features (indexes, GetKey, etc.); I think those should have been separate followup PRs, and I'm going to push back harder against similar large PRs in the future.

I've left a bunch of comments about uses of unwraps in the new backend—we know this is a problem with the current one, and I'd like to try to start out building it right on our second try.

arihant2math commented on components/script/dom/idbobjectstore.rs:
```diff
@@ -125,16 +126,50 @@ impl IDBObjectStore {
 
     /// Checks if the transation is active, throwing a "TransactionInactiveError" DOMException if not.
-    fn check_transaction_active(&self) -> Fallible<()> {
+    fn check_transaction_active(&self) -> Fallible<DomRoot<GlobalScope>> {
```
Yeah it was an hacky optimization, I needed the global scope for creating a channel (see `create_channel`). Merging that PR could help.

arihant2math replied:
Something like `get_global_scope` or something.

jdm replied:
It should also be the same as `self.global()`.

arihant2math commented on components/script/dom/idbobjectstore.rs:
```diff
@@ -125,16 +126,50 @@ impl IDBObjectStore {
 
-        receiver.recv().unwrap()
+        // First unwrap for ipc
+        // Second unwrap will never happen unless this db gets manually deleted somehow
+        receiver.recv().unwrap().unwrap()
```
The second unwrap is needed. I think the main issue here is that the backend has no state, so the request could be about a db that no longer exists. I think we could just error out of the function?

arihant2math commented on components/net/indexeddb/engines/sqlite/serialize.rs:
```diff
@@ -0,0 +1,360 @@
+
+// Tests for the codec module
+#[cfg(test)]
+mod tests {
```
I'm just going to remove this module from this merge request and keep using bincode instead.

It's an approach that leaves key ranges broken, but it should easily resolvable in a separate pr.

arihant2math commented:
I think the messaging system will need another rework later on. Representing every single return state is terribly verbose to handle.

jdm reviewed (approved):
Thanks!

mrobinson reviewed (commented):
Shouldn't this remove the `heed` dependency as well?

arihant2math commented:
> Shouldn't this remove the `heed` dependency as well?

Good point, removed!

arihant2math commented:
Also fixed the singular wpt test.

janvarga commented:
> > Binary size increase of 5.4 mb on aarch64 macos (release profile).
> 
> This is a pretty significant increase unfortunately. Is it possible to use a lower-level sqlite crate that doesn't bring in so many dependencies? Maybe it's worth discussing with a wider group of people in the community to figure out the pros and cons.

Yeah.

Just so you know, the stuff I'm experimenting with should only require rusqlite in future, which is probably inevitable anyway if we want to support various storage-related features.

mrobinson commented:
Here's the relevant bencher results: https://bencher.dev/perf/servo?report=b229b5d8-3ce3-481e-961c-382fa5aa6be2&branches=8df49e2b-ab2b-4185-9b3b-da59b0463338&heads=d741596e-1120-4712-8d3f-cf6b8afee69b&testbeds=d742c702-3842-4108-9d0c-2db74e57599a&benchmarks=1373a43f-100f-4ade-9524-c9c2f3acd487&measures=678e4118-c8a5-494d-8799-08abc3021cd5&start_time=1751281324000&end_time=1753873324000&key=true&reports_per_page=4&branches_per_page=8&testbeds_per_page=8&benchmarks_per_page=8&plots_per_page=8&reports_page=1&branches_page=1&testbeds_page=1&benchmarks_page=1&plots_page=1

arihant2math commented:
> Just so you know, the stuff I'm experimenting with should only require rusqlite in future, which is probably inevitable anyway if we want to support various storage-related features.

Agreed, rusqlite is the way forward

arihant2math commented:
> > Binary size increase of 5.4 mb on aarch64 macos (release profile).
> 
> This is a pretty significant increase unfortunately. Is it possible to use a lower-level sqlite crate that doesn't bring in so many dependencies? Maybe it's worth discussing with a wider group of people in the community to figure out the pros and cons.

Ok. I'll move down one level of abstraction to sea-query.
sea-query would integrate directly with rusqlite, so I'd hit two birds with one stone I suppose. However it'll increase verbosity by quite a bit.

arihant2math commented:
> Is sea-query requiring that you keep doing these `HANDLE.block_on(async { ... })` calls to ultimately call the synchronous rusqlite apis?
no, I'm using a rusqlite connection to manage everything anyhow, so no async will be used in that regard.

arihant2math commented:
My preliminary port to rusqlite is not showing any binary size improvements, I still do have to investigate further however.

arihant2math commented:
... and all the wpt tests are failing again

arihant2math commented:
Weird, any clue why that happens?

mrobinson commented:
Looks like this change makes the build require SQLite shared libraries in the environment:

```
z,now" "-Wl,-O1" "-Wl,--strip-debug" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: /usr/bin/ld: cannot find -lsqlite3: No such file or directory
          collect2: error: ld returned 1 exit status
         
```

arihant2math commented:
Ok. I've vendored sqlite3.

arihant2math commented:
Try run passes.

arihant2math commented:
> ✨ Try run ([#16749035886](https://github.com/servo/servo/actions/runs/16749035886)) succeeded.

1.95 mb increase.

janvarga commented:
So sqlite3 will be statically linked? That’s not great for build size, but on the other hand, it can help avoid a lot of issues, since system sqlite3 is often built with different options.

jdm commented on components/net/indexeddb/idb_thread.rs:
```diff
@@ -280,19 +302,17 @@ impl IndexedDBManager {
                     },
                     Entry::Occupied(db) => {
-                        let _ = sender.send(db.get().version);
+                        let _ = sender.send(db.get().version().unwrap_or(version.unwrap_or(0)));
```
We repeat `version.unwrap_or(0)` several times in this function; can we use store that in a variable?

jdm commented on components/net/indexeddb/idb_thread.rs:
```diff
@@ -304,17 +324,8 @@ impl IndexedDBManager {
-                if std::fs::remove_dir_all(&db_dir).is_err() {
-                    let _ = sender.send(Err(()));
+                if let Some(db) = self.databases.remove(&idb_description) {
+                    db.delete_database(sender);
```
Don't we want to send an `Ok(())` value in this branch as well? Do we need the `else` at all?

jdm reviewed (commented):
I reviewed everything but engines/sqlite.rs before I ran out of time. I'll try to get back to this by tomorrow.

jdm reviewed (changes requested):
Looking pretty close to mergeable!

arihant2math committed:
address review

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

arihant2math committed:
address review

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

arihant2math commented on components/net/indexeddb/engines/sqlite.rs:
```diff
@@ -0,0 +1,688 @@
+        let _ = self.connection.close();
+        if self.db_path.exists() {
+            if let Err(e) = std::fs::remove_dir_all(self.db_path.parent().unwrap()) {
+                error!("Failed to delete database: {:?}", e);
```
I think this function is highly unlikely to fail, so just logging errors is good enough for debugging. Imo erroring out would just push the issue one level higher and also require a custom error type.

jdm commented:
This can merge once `./mach test-tidy` passes.
